<a href="https://colab.research.google.com/github/Neethu0207/Recommend-System/blob/main/SkinRecum.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing libraries**

In [ ]:

import tensorflow as tf
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import plotly.express as px
import plotly.graph_objects as go
import plotly.offline as pyo
import plotly.io as pio
import seaborn as sns
import missingno as mno
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD
from sklearn.manifold import TSNE

from bokeh.io import curdoc, push_notebook, output_notebook
from bokeh.layouts import column, layout
from bokeh.models import ColumnDataSource, Div, Select, Slider, TextInput, HoverTool
from bokeh.plotting import figure, show


**Loading the dataset**

In [ ]:
df=pd.read_csv('/content/MP-Skin Care Product Recommendation System3.csv')

**Data Exploration**

In [ ]:
df.head()

product_href  \
0  https://www.beautyhaul.com/product/detail/bubb...   
1  https://www.sociolla.com/face-wash/62452-ph-ba...   
2  https://www.sociolla.com/toner/15871-licorice-...   
3  https://www.beautyhaul.com/product/detail/aqua...   
4  https://www.sociolla.com/essence/38023-licoric...   

                                    product_name product_type    brand  \
0       ACWELL Bubble Free PH Balancing Cleanser    Face Wash  ACWELL    
1    ACWELL pH Balancing Soothing Cleansing Foam    Face Wash  ACWELL    
2  Acwell Licorice pH Balancing Cleansing Toner         Toner  ACWELL    
3                ACWELL Aquaseal Soothing Tonic         Toner  ACWELL    
4             Licorice pH Balancing Essence Mist        Toner  ACWELL    

                                 notable_effects  \
0  Acne-Free, Pore-Care, Brightening, Anti-Aging   
1                            Soothing, Balancing   
2                            Soothing, Balancing   
3  Acne-Free, Pore-Care, Brightening, Anti-Aging   
4                          Brightening, Soothing   

                                    skintype       price  \
0                                       Oily  Rp 209.000   
1                   Normal, Dry, Combination  Rp 181.800   
2  Normal, Dry, Oily, Combination, Sensitive  Rp 149.000   
3                                       Oily  Rp 290.000   
4                                Normal, Dry  Rp 194.650   

                                         description  \
0  Mengangkat kotoran dan menghapus makeup dalam ...   
1  Membersihkan dan menenangkan kulit sensitif de...   
2  Mengangkat sisa kotoran, debu, dan make up sek...   
3  Pre-essence yang diformulasikan dengan ekstrak...   
4  Essens mist dengan kelembapan tinggi yang memb...   

                                         picture_src  
0  https://www.beautyhaul.com/assets/uploads/prod...  
1  https://images.soco.id/8f08ced0-344d-41f4-a15e...  
2  https://www.soco.id/cdn-cgi/image/w=73,format=...  
3  https://www.beautyhaul.com/assets/uploads/prod...  
4  https://www.sociolla.com/cdn-cgi/image/w=425,f...

In [ ]:
df.shape

(1224, 9)

In [ ]:
df.size

11016

In [ ]:
df.dtypes

product_href       object
product_name       object
product_type       object
brand              object
notable_effects    object
skintype           object
price              object
description        object
picture_src        object
dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1224 entries, 0 to 1223
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   product_href     1224 non-null   object
 1   product_name     1224 non-null   object
 2   product_type     1224 non-null   object
 3   brand            1224 non-null   object
 4   notable_effects  1224 non-null   object
 5   skintype         1224 non-null   object
 6   price            1224 non-null   object
 7   description      1224 non-null   object
 8   picture_src      1224 non-null   object
dtypes: object(9)
memory usage: 86.2+ KB


In [ ]:
df.isnull().sum()

product_href       0
product_name       0
product_type       0
brand              0
notable_effects    0
skintype           0
price              0
description        0
picture_src        0
dtype: int64

In [ ]:

df.duplicated().sum()

np.int64(14)

In [ ]:
pro = df[['brand','product_name','product_type','notable_effects','price']]
pro.describe()

brand                          product_name product_type  \
count        1224                                  1224         1224   
unique        211                                  1192            5   
top     SOMETHINC  Somethinc RESURRECT Multibiome Serum        Serum   
freq           80                                     4          307   

                           notable_effects      price  
count                                 1224       1224  
unique                                 151        512  
top     Pore-Care, Brightening, Anti-Aging  Rp 89.000  
freq                                   150         31

**From the above info**
1. In total, there are 1224 products scrapped from websites.
2. Unfortunately, there are 14 duplicate rows. Need to be removed.
Of 5 types of products, serum is more hype than others.
3. SOMETHINC is the top brand, which means they have many kinds of products.
4. From many pairs of notable effects, 150 products are good for pore care, brightening, and anti-aging all in one product.
5. Looks like many skin care products suitable for oily skin.

In [ ]:
count_brand=df['brand'].value_counts()
count_percentage=df['brand'].value_counts(1)*100

count_dfbrand_=pd.DataFrame({'Brand':count_brand.index,'Counts':count_brand.values,'Percent%':np.round(count_percentage.values,2)})
top_10_brands = count_dfbrand_.head(10)
top_10_brands


Brand  Counts  Percent%
0        SOMETHINC      80      6.54
1           WARDAH      69      5.64
2        INNISFREE      49      4.00
3          AVOSKIN      39      3.19
4             ERHA      33      2.70
5       ELSHE SKIN      24      1.96
6    THE BODY SHOP      23      1.88
7      BANANA BOAT      22      1.80
8          AZARINE      20      1.63
9  NATURE REPUBLIC      18      1.47

In [ ]:
count_brand=df['brand'].value_counts()
count_percentage=df['brand'].value_counts()*100

count_dfbrand_=pd.DataFrame({'Brand':count_brand.index,'Counts':count_brand.values,'Percent%':np.round(count_percentage.values,2)})
top_10_brands = count_dfbrand_.head(10)
top_10_brands


Brand  Counts  Percent%
0        SOMETHINC      80      8000
1           WARDAH      69      6900
2        INNISFREE      49      4900
3          AVOSKIN      39      3900
4             ERHA      33      3300
5       ELSHE SKIN      24      2400
6    THE BODY SHOP      23      2300
7      BANANA BOAT      22      2200
8          AZARINE      20      2000
9  NATURE REPUBLIC      18      1800

In [ ]:
#visualising too 10 brands
fig = px.bar(top_10_brands, x='Brand', y='Counts', color='Brand',
             color_discrete_sequence=px.colors.qualitative.Plotly,
             title='Top 10 Brands')
fig.show()

In [ ]:
#checking product type
count_product=df['product_type'].value_counts()
print(count_product)

product_type
Serum          307
Toner          255
Moisturizer    248
Sunscreen      212
Face Wash      202
Name: count, dtype: int64


In [ ]:
#calculating product type percentage
count_percentage=df['product_type'].value_counts(1)*100
print(count_percentage)

product_type
Serum          25.081699
Toner          20.833333
Moisturizer    20.261438
Sunscreen      17.320261
Face Wash      16.503268
Name: proportion, dtype: float64


In [ ]:
#visualising product type using piechart
fig = px.pie(df, names='product_type', title='Product Type Distribution')
fig.show()

In [ ]:
#visualsing procuct type using histogram
fig = px.histogram(df, x='product_type', title='Product Type Distribution')
fig.show()

In [ ]:
#calculating product name
count_productnam=df['product_name'].value_counts()
print(count_productnam)

product_name
Somethinc RESURRECT Multibiome Serum                   4
Somethinc Skin Defender Bakuchiol + COQ10 Serum        4
Somethinc 5% Niacinamide Barrier Serum                 3
Dear Me Beauty Watermelon Multipurpose Gel             3
Innisfree Retinol Cica Moisture Recovery Serum 30ml    3
                                                      ..
GENTLE HOUR Dew It Again Hydrating Toner               1
GENTLE HOUR Brighter Days Exfoliating Toner            1
Gentle Hour Clean Start Hydrating Jelly Cleanser       1
Garnier Sakura Night Cream                             1
GLAMGLOW YOUTHPOTION Rejuvinating Peptida Serum        1
Name: count, Length: 1192, dtype: int64


In [ ]:
#calculating no of counts of each product name
count_productnam=df['product_name'].value_counts(1)*100
print(count_productnam)

product_name
Somethinc RESURRECT Multibiome Serum                   0.326797
Somethinc Skin Defender Bakuchiol + COQ10 Serum        0.326797
Somethinc 5% Niacinamide Barrier Serum                 0.245098
Dear Me Beauty Watermelon Multipurpose Gel             0.245098
Innisfree Retinol Cica Moisture Recovery Serum 30ml    0.245098
                                                         ...   
GENTLE HOUR Dew It Again Hydrating Toner               0.081699
GENTLE HOUR Brighter Days Exfoliating Toner            0.081699
Gentle Hour Clean Start Hydrating Jelly Cleanser       0.081699
Garnier Sakura Night Cream                             0.081699
GLAMGLOW YOUTHPOTION Rejuvinating Peptida Serum        0.081699
Name: proportion, Length: 1192, dtype: float64


In [ ]:
#selecting top 5 product names used
top_5_product_names = count_productnam.head(5)
print(top_5_product_names)

product_name
Somethinc RESURRECT Multibiome Serum                   0.326797
Somethinc Skin Defender Bakuchiol + COQ10 Serum        0.326797
Somethinc 5% Niacinamide Barrier Serum                 0.245098
Dear Me Beauty Watermelon Multipurpose Gel             0.245098
Innisfree Retinol Cica Moisture Recovery Serum 30ml    0.245098
Name: proportion, dtype: float64


In [ ]:
#skin type that suitable for the product
count=df['skintype'].value_counts()
count_percentage=df['skintype'].value_counts(1)*100
count_df=pd.DataFrame({'Skin Type':count.index,'Counts':count.values,'Percent%':np.round(count_percentage.values,2)})
count_df

Skin Type  Counts  Percent%
0                                        Oily     310     25.33
1   Normal, Dry, Oily, Combination, Sensitive     263     21.49
2                                         Dry     167     13.64
3                                 Normal, Dry      93      7.60
4              Normal, Dry, Oily, Combination      68      5.56
5                                   Sensitive      55      4.49
6                Normal, Dry, Oily, Sensitive      52      4.25
7                Oily, Combination, Sensitive      43      3.51
8                        Dry, Oily, Sensitive      36      2.94
9                              Dry, Sensitive      27      2.21
10                   Normal, Dry, Combination      24      1.96
11                          Oily, Combination      24      1.96
12                            Oily, Sensitive      24      1.96
13                        Normal, Combination      19      1.55
14                               Normal, Oily      19      1.55

In [ ]:
#visualising using piechart
fig = px.pie(df, names='skintype', title='Skin Type Distribution')
fig.show()

In [ ]:
# Creating numerical labels of notable_effects due to many effects shown by the diagram above
df['notable_effects']=df['notable_effects'].astype('category')
df['notable_effects_label']=df['notable_effects'].cat.codes
df.head()

product_href  \
0  https://www.beautyhaul.com/product/detail/bubb...   
1  https://www.sociolla.com/face-wash/62452-ph-ba...   
2  https://www.sociolla.com/toner/15871-licorice-...   
3  https://www.beautyhaul.com/product/detail/aqua...   
4  https://www.sociolla.com/essence/38023-licoric...   

                                    product_name product_type    brand  \
0       ACWELL Bubble Free PH Balancing Cleanser    Face Wash  ACWELL    
1    ACWELL pH Balancing Soothing Cleansing Foam    Face Wash  ACWELL    
2  Acwell Licorice pH Balancing Cleansing Toner         Toner  ACWELL    
3                ACWELL Aquaseal Soothing Tonic         Toner  ACWELL    
4             Licorice pH Balancing Essence Mist        Toner  ACWELL    

                                 notable_effects  \
0  Acne-Free, Pore-Care, Brightening, Anti-Aging   
1                            Soothing, Balancing   
2                            Soothing, Balancing   
3  Acne-Free, Pore-Care, Brightening, Anti-Aging   
4                          Brightening, Soothing   

                                    skintype       price  \
0                                       Oily  Rp 209.000   
1                   Normal, Dry, Combination  Rp 181.800   
2  Normal, Dry, Oily, Combination, Sensitive  Rp 149.000   
3                                       Oily  Rp 290.000   
4                                Normal, Dry  Rp 194.650   

                                         description  \
0  Mengangkat kotoran dan menghapus makeup dalam ...   
1  Membersihkan dan menenangkan kulit sensitif de...   
2  Mengangkat sisa kotoran, debu, dan make up sek...   
3  Pre-essence yang diformulasikan dengan ekstrak...   
4  Essens mist dengan kelembapan tinggi yang memb...   

                                         picture_src  notable_effects_label  
0  https://www.beautyhaul.com/assets/uploads/prod...                     18  
1  https://images.soco.id/8f08ced0-344d-41f4-a15e...                    127  
2  https://www.soco.id/cdn-cgi/image/w=73,format=...                    127  
3  https://www.beautyhaul.com/assets/uploads/prod...                     18  
4  https://www.sociolla.com/cdn-cgi/image/w=425,f...                     40

In [ ]:
#skin that siutable for
count=df['notable_effects'].value_counts()
count_percentage=df['notable_effects'].value_counts(1)*100
count_df=pd.DataFrame({'notable_effects':count.index,'Counts':count.values,'Percent%':np.round(count_percentage.values,2)})
count_df

notable_effects  Counts  Percent%
0       Pore-Care, Brightening, Anti-Aging     150     12.25
1                               Anti-Aging     105      8.58
2        Acne-Free, Oil-Control, Pore-Care      86      7.03
3    Moisturizing, Brightening, Black-Spot      71      5.80
4                  Hydrating, Moisturizing      70      5.72
..                                     ...     ...       ...
146     Soothing, Moisturizing, Anti-Aging       1      0.08
147      Soothing, Refreshing, Oil-Control       1      0.08
148                   Soothing, Refreshing       1      0.08
149       Soothing, Oil-Control, Pore-Care       1      0.08
150           UV-Protection, No-Whitecast        1      0.08

[151 rows x 3 columns]

In [ ]:
#visualising top  notable_effects using histogram
fig = px.histogram(df, x='notable_effects', title='Notable Effects Distribution')
fig.show()

In [ ]:
#visualising only top 5 notable effects using barplot
top_5_effects = count_df.head(5)
fig = px.bar(top_5_effects, x='notable_effects', y='Counts', color='notable_effects',
             color_discrete_sequence=px.colors.qualitative.Plotly,
             title='Top 5 Notable Effects')
fig

Data Preparation

In [ ]:
# Data Preparation
df = pd.read_csv('/content/MP-Skin Care Product Recommendation System3.csv')

from sklearn.preprocessing import MultiLabelBinarizer
import pandas as pd

if 'skintype' in df.columns:
    df['skintype'] = df['skintype'].astype(str).fillna('').str.split(',')
    df['skintype'] = df['skintype'].apply(lambda x: [item.strip() for item in x if item.strip()])

    mlb = MultiLabelBinarizer()
    skintype_encoded = pd.DataFrame(mlb.fit_transform(df['skintype']),
                                     columns=mlb.classes_,
                                     index=df.index)
    df = df.drop('skintype', axis=1).join(skintype_encoded)


    print("DataFrame head after skintype encoding:")
    display(df.head())

else:
    print("Skipping skintype encoding: 'skintype' column not found.")
    print("Current DataFrame head:")
    display(df.head())

DataFrame head after skintype encoding:


product_href  \
0  https://www.beautyhaul.com/product/detail/bubb...   
1  https://www.sociolla.com/face-wash/62452-ph-ba...   
2  https://www.sociolla.com/toner/15871-licorice-...   
3  https://www.beautyhaul.com/product/detail/aqua...   
4  https://www.sociolla.com/essence/38023-licoric...   

                                    product_name product_type    brand  \
0       ACWELL Bubble Free PH Balancing Cleanser    Face Wash  ACWELL    
1    ACWELL pH Balancing Soothing Cleansing Foam    Face Wash  ACWELL    
2  Acwell Licorice pH Balancing Cleansing Toner         Toner  ACWELL    
3                ACWELL Aquaseal Soothing Tonic         Toner  ACWELL    
4             Licorice pH Balancing Essence Mist        Toner  ACWELL    

                                 notable_effects       price  \
0  Acne-Free, Pore-Care, Brightening, Anti-Aging  Rp 209.000   
1                            Soothing, Balancing  Rp 181.800   
2                            Soothing, Balancing  Rp 149.000   
3  Acne-Free, Pore-Care, Brightening, Anti-Aging  Rp 290.000   
4                          Brightening, Soothing  Rp 194.650   

                                         description  \
0  Mengangkat kotoran dan menghapus makeup dalam ...   
1  Membersihkan dan menenangkan kulit sensitif de...   
2  Mengangkat sisa kotoran, debu, dan make up sek...   
3  Pre-essence yang diformulasikan dengan ekstrak...   
4  Essens mist dengan kelembapan tinggi yang memb...   

                                         picture_src  Combination  Dry  \
0  https://www.beautyhaul.com/assets/uploads/prod...            0    0   
1  https://images.soco.id/8f08ced0-344d-41f4-a15e...            1    1   
2  https://www.soco.id/cdn-cgi/image/w=73,format=...            1    1   
3  https://www.beautyhaul.com/assets/uploads/prod...            0    0   
4  https://www.sociolla.com/cdn-cgi/image/w=425,f...            0    1   

   Normal  Oily  Sensitive  
0       0     1          0  
1       1     0          0  
2       1     1          1  
3       0     1          0  
4       1     0          0

In [ ]:
#saving into csv file
df.to_csv('MP-Skin Care Product Recommendation System3.csv', index=False)

# TF-IDF

The TF-IDF Vectorizer will be used in the recommendation system to find a representation of the important features of each notable_effects category. We will use the tfidfvectorizer() function from the sklearn library.

In [ ]:
#initialzing tf-idf
tfidf = TfidfVectorizer(stop_words='english')

In [ ]:
#perform idf on notable_effects
tfidf_matrix = tfidf.fit_transform(df['notable_effects'])

In [ ]:
#getting features name out
tfidf.get_feature_names_out()

array(['acne', 'aging', 'anti', 'balancing', 'barrier', 'black',
       'brightening', 'care', 'control', 'free', 'hydrating',
       'moisturizing', 'oil', 'pore', 'protection', 'refreshing', 'skin',
       'soothing', 'spot', 'uv', 'whitecast'], dtype=object)

In [ ]:
#initialzing tf-idf
tfidf = TfidfVectorizer(stop_words='english')

#perform idf on notable_effects
tfidf_matrix = tfidf.fit_transform(df['notable_effects'])

#getting features name out
tfidf.get_feature_names_out()

#converting into matrix format
# The user mistakenly used the alias 'tf' (tensorflow) instead of the 'tfidf' variable.
# Correcting this to use the initialized TfidfVectorizer.
tfidf_matrix = tfidf.fit_transform(df['notable_effects'])

In [ ]:
#viewing matrix size
shape=tfidf_matrix.shape
print(shape)

(1224, 21)


In [ ]:
#convert tfidf into matrix
tfidf_matrix.toarray()

array([[0.44842197, 0.32844212, 0.32844212, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.41403597, 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.42500873, 0.42500873, ..., 0.        , 0.        ,
        0.        ]])

In [ ]:

pd.DataFrame(
    tfidf_matrix.toarray(),
    columns=tfidf.get_feature_names_out(),
    index=df['product_name']
).sample(shape[1],axis=1).sample(10,axis=0)

pore  soothing  \
product_name                                                             
COSRX Hydrium Watery Toner                          0.000000  0.000000   
HADA LABO Tamagohada Mild Peeling Face Wash 50gr    0.325576  0.000000   
Dear Klairs Freshly Juiced Vitamin E Mask           0.000000  0.000000   
HAPLE Moonlight Glow (toner AHA)                    0.481907  0.000000   
Somethinc SOS Bakuchiol Electrolyte Rich Moistu...  0.530327  0.000000   
Mediheal Tea Tree Biome Blemish Cica Cream          0.372413  0.000000   
Wardah Acnederm Pure Foaming Cleanser               0.451557  0.000000   
SCARLETT WHITENING Brightly Ever After Serum        0.000000  0.000000   
L'ESSENTIAL Triple Cream                            0.481907  0.000000   
SAFI Dermasafe Rehydrate & Recover Night Moistu...  0.000000  0.433013   

                                                        free      care  \
product_name                                                             
COSRX Hydrium Watery Toner                          0.528341  0.000000   
HADA LABO Tamagohada Mild Peeling Face Wash 50gr    0.392642  0.325576   
Dear Klairs Freshly Juiced Vitamin E Mask           0.000000  0.000000   
HAPLE Moonlight Glow (toner AHA)                    0.000000  0.481907   
Somethinc SOS Bakuchiol Electrolyte Rich Moistu...  0.000000  0.530327   
Mediheal Tea Tree Biome Blemish Cica Cream          0.449127  0.372413   
Wardah Acnederm Pure Foaming Cleanser               0.544574  0.451557   
SCARLETT WHITENING Brightly Ever After Serum        0.000000  0.000000   
L'ESSENTIAL Triple Cream                            0.000000  0.481907   
SAFI Dermasafe Rehydrate & Recover Night Moistu...  0.000000  0.000000   

                                                        skin  refreshing  \
product_name                                                               
COSRX Hydrium Watery Toner                          0.000000         0.0   
HADA LABO Tamagohada Mild Peeling Face Wash 50gr    0.000000         0.0   
Dear Klairs Freshly Juiced Vitamin E Mask           0.000000         0.0   
HAPLE Moonlight Glow (toner AHA)                    0.000000         0.0   
Somethinc SOS Bakuchiol Electrolyte Rich Moistu...  0.000000         0.0   
Mediheal Tea Tree Biome Blemish Cica Cream          0.000000         0.0   
Wardah Acnederm Pure Foaming Cleanser               0.000000         0.0   
SCARLETT WHITENING Brightly Ever After Serum        0.000000         0.0   
L'ESSENTIAL Triple Cream                            0.000000         0.0   
SAFI Dermasafe Rehydrate & Recover Night Moistu...  0.637377         0.0   

                                                    whitecast  black  \
product_name                                                           
COSRX Hydrium Watery Toner                                0.0    0.0   
HADA LABO Tamagohada Mild Peeling Face Wash 50gr          0.0    0.0   
Dear Klairs Freshly Juiced Vitamin E Mask                 0.0    0.0   
HAPLE Moonlight Glow (toner AHA)                          0.0    0.0   
Somethinc SOS Bakuchiol Electrolyte Rich Moistu...        0.0    0.0   
Mediheal Tea Tree Biome Blemish Cica Cream                0.0    0.0   
Wardah Acnederm Pure Foaming Cleanser                     0.0    0.0   
SCARLETT WHITENING Brightly Ever After Serum              0.0    0.0   
L'ESSENTIAL Triple Cream                                  0.0    0.0   
SAFI Dermasafe Rehydrate & Recover Night Moistu...        0.0    0.0   

                                                    hydrating  spot  ...  \
product_name                                                         ...   
COSRX Hydrium Watery Toner                                0.0   0.0  ...   
HADA LABO Tamagohada Mild Peeling Face Wash 50gr          0.0   0.0  ...   
Dear Klairs Freshly Juiced Vitamin E Mask                 0.0   0.0  ...   
HAPLE Moonlight Glow (toner AHA)                          0.0   0.0  ...   
Somethinc SOS Bakuchiol Electrolyte Rich Moi

In [ ]:
#making dataframe to see tfidf matrix
df_tfidf = pd.DataFrame(tfidf_matrix.toarray(), columns=tfidf.get_feature_names_out())
df_tfidf

acne     aging      anti  balancing  barrier  black  brightening  \
0     0.448422  0.328442  0.328442   0.000000      0.0    0.0     0.322602   
1     0.000000  0.000000  0.000000   0.820405      0.0    0.0     0.000000   
2     0.000000  0.000000  0.000000   0.820405      0.0    0.0     0.000000   
3     0.448422  0.328442  0.328442   0.000000      0.0    0.0     0.322602   
4     0.000000  0.000000  0.000000   0.000000      0.0    0.0     0.559172   
...        ...       ...       ...        ...      ...    ...          ...   
1219  0.000000  0.000000  0.000000   0.000000      0.0    0.0     0.426141   
1220  0.000000  0.000000  0.000000   0.000000      0.0    0.0     0.000000   
1221  0.414036  0.000000  0.000000   0.000000      0.0    0.0     0.000000   
1222  0.000000  0.000000  0.000000   0.000000      0.0    0.0     0.000000   
1223  0.000000  0.425009  0.425009   0.000000      0.0    0.0     0.417452   

          care  control      free  ...  moisturizing  oil      pore  \
0     0.372413      0.0  0.449127  ...      0.000000  0.0  0.372413   
1     0.000000      0.0  0.000000  ...      0.000000  0.0  0.000000   
2     0.000000      0.0  0.000000  ...      0.000000  0.0  0.000000   
3     0.372413      0.0  0.449127  ...      0.000000  0.0  0.372413   
4     0.000000      0.0  0.000000  ...      0.000000  0.0  0.000000   
...        ...      ...       ...  ...           ...  ...       ...   
1219  0.000000      0.0  0.000000  ...      0.000000  0.0  0.000000   
1220  0.000000      0.0  0.000000  ...      0.421956  0.0  0.000000   
1221  0.000000      0.0  0.414687  ...      0.000000  0.0  0.000000   
1222  0.000000      0.0  0.000000  ...      0.630729  0.0  0.000000   
1223  0.481907      0.0  0.000000  ...      0.000000  0.0  0.481907   

      protection  refreshing  skin  soothing  spot        uv  whitecast  
0       0.000000    0.000000   0.0  0.000000   0.0  0.000000        0.0  
1       0.000000    0.000000   0.0  0.571783   0.0  0.000000        0.0  
2       0.000000    0.000000   0.0  0.571783   0.0  0.000000        0.0  
3       0.000000    0.000000   0.0  0.000000   0.0  0.000000        0.0  
4       0.000000    0.000000   0.0  0.829052   0.0  0.000000        0.0  
...          ...         ...   ...       ...   ...       ...        ...  
1219    0.639689    0.000000   0.0  0.000000   0.0  0.639689        0.0  
1220    0.000000    0.906616   0.0  0.000000   0.0  0.000000        0.0  
1221    0.000000    0.810314   0.0  0.000000   0.0  0.000000        0.0  
1222    0.000000    0.000000   0.0  0.000000   0.0  0.000000        0.0  
1223    0.000000    0.000000   0.0  0.000000   0.0  0.000000        0.0  

[1224 rows x 21 columns]

In [ ]:
#calculating coisine similarity
from sklearn.metrics.pairwise import cosine_similarity
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)
cosine_sim

array([[1.        , 0.        , 0.        , ..., 0.37190959, 0.        ,
        0.77278913],
       [0.        , 1.        , 1.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 1.        , 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.37190959, 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.77278913, 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:

# Creating a dataframe from the cosine_similarity variable with rows and columns in the form of product names
cosine_sim_df = pd.DataFrame(cosine_sim, index=df['product_name'], columns=df['product_name'])
cosine_sim_df

product_name                                        ACWELL Bubble Free PH Balancing Cleanser  \
product_name                                                                                   
ACWELL Bubble Free PH Balancing Cleanser                                            1.000000   
ACWELL pH Balancing Soothing Cleansing Foam                                         0.000000   
Acwell Licorice pH Balancing Cleansing Toner                                        0.000000   
ACWELL Aquaseal Soothing Tonic                                                      1.000000   
Licorice pH Balancing Essence Mist                                                  0.180390   
...                                                                                      ...   
Whitelab Brightening Day Cream                                                      0.137474   
YOU The Radiance White Purifying Facial Foam                                        0.000000   
YOU Basic Skin Care Simply Fresh and Bright Fac...                                  0.371910   
YOU Dazzling Glow Up Facial Foam                                                    0.000000   
YOU Golden Age Deep Cleansing Facial Wash                                           0.772789   

product_name                                        ACWELL pH Balancing Soothing Cleansing Foam  \
product_name                                                                                      
ACWELL Bubble Free PH Balancing Cleanser                                               0.000000   
ACWELL pH Balancing Soothing Cleansing Foam                                            1.000000   
Acwell Licorice pH Balancing Cleansing Toner                                           1.000000   
ACWELL Aquaseal Soothing Tonic                                                         0.000000   
Licorice pH Balancing Essence Mist                                                     0.474038   
...                                                                                         ...   
Whitelab Brightening Day Cream                                                         0.000000   
YOU The Radiance White Purifying Facial Foam                                           0.000000   
YOU Basic Skin Care Simply Fresh and Bright Fac...                                     0.000000   
YOU Dazzling Glow Up Facial Foam                                                       0.000000   
YOU Golden Age Deep Cleansing Facial Wash                                              0.000000   

product_name                                        Acwell Licorice pH Balancing Cleansing Toner   \
product_name                                                                                        
ACWELL Bubble Free PH Balancing Cleanser                                                 0.000000   
ACWELL pH Balancing Soothing Cleansing Foam                                              1.000000   
Acwell Licorice pH Balancing Cleansing Toner                                             1.000000   
ACWELL Aquaseal Soothing Tonic                                                           0.000000   
Licorice pH Balancing Essence Mist                                                       0.474038   
...                                                                                           ...   
Whitelab Brightening Day Cream                                                           0.000000   
YOU The Radiance White Purifying Facial Foam                                             0.000000   
YOU Basic Skin Care Simply Fresh and Bright Fac...                                       0.000000   
YOU Dazzling Glow Up Facial Foam                                                         0.000000   
YOU Golden Age Deep Cleansing Facial Wash                                                0.000000   

product_name                                        ACWELL Aquaseal Soothing Tonic   \
product_name                                                  

In [ ]:
#checking similarity with each product
cosine_sim_df.sample(5, axis=1).sample(10, axis=0)


product_name                                        Everwhite Acnepair Toner   \
product_name                                                                    
BIODERMA Photoderm Max fluide SPF100 40ml                            0.000000   
Mediheal Tea Tree Biome Blemish Cica Cream                           0.594637   
Innisfree Soybean Energy Skin EX 200ml                               0.000000   
Whitelab Brightening Facial Wash                                     0.594637   
PURIVERA BOTANICALS Sugar Willow Serum - Fruit ...                   0.496176   
SIMPLISTIC Cleanser begin                                            0.333120   
SUKIN Signature Original Hydrating Mist Toner                        0.000000   
Innisfree Jeju Cherry Blossom Mist 120 mL                            0.000000   
AVOSKIN YOUR SKIN BAE SERIES Panthenol 5% + Mug...                   0.000000   
KOSE Sekkisei White UV Milk SPF 50+ PA++++                           0.000000   

product_name                                        BIO BEAUTY LAB Acne Treatment Face Oil  \
product_name                                                                                 
BIODERMA Photoderm Max fluide SPF100 40ml                                         0.000000   
Mediheal Tea Tree Biome Blemish Cica Cream                                        0.301309   
Innisfree Soybean Energy Skin EX 200ml                                            0.000000   
Whitelab Brightening Facial Wash                                                  0.301309   
PURIVERA BOTANICALS Sugar Willow Serum - Fruit ...                                0.723345   
SIMPLISTIC Cleanser begin                                                         0.000000   
SUKIN Signature Original Hydrating Mist Toner                                     0.000000   
Innisfree Jeju Cherry Blossom Mist 120 mL                                         0.000000   
AVOSKIN YOUR SKIN BAE SERIES Panthenol 5% + Mug...                                0.000000   
KOSE Sekkisei White UV Milk SPF 50+ PA++++                                        0.000000   

product_name                                        Mamonde Everyday Sun Stick 50+++++  \
product_name                                                                             
BIODERMA Photoderm Max fluide SPF100 40ml                                     0.827608   
Mediheal Tea Tree Biome Blemish Cica Cream                                    0.000000   
Innisfree Soybean Energy Skin EX 200ml                                        0.000000   
Whitelab Brightening Facial Wash                                              0.000000   
PURIVERA BOTANICALS Sugar Willow Serum - Fruit ...                            0.000000   
SIMPLISTIC Cleanser begin                                                     0.000000   
SUKIN Signature Original Hydrating Mist Toner                                 0.000000   
Innisfree Jeju Cherry Blossom Mist 120 mL                                     0.000000   
AVOSKIN YOUR SKIN BAE SERIES Panthenol 5% + Mug...                            0.000000   
KOSE Sekkisei White UV Milk SPF 50+ PA++++                                    0.858849   

product_name                                        Lagom Cellus Sun Gel  \
product_name                                                               
BIODERMA Photoderm Max fluide SPF100 40ml                       0.530562   
Mediheal Tea Tree Biome Blemish Cica Cream                      0.000000   
Innisfree Soybean Energy Skin EX 200ml                          0.000000   
Whitelab Brightening Facial Wash                                0.000000   
PURIVERA BOTANICALS Sugar Willow Serum - Fruit ...              0.000000   
SIMPLISTIC Cleanser begin                                       0.000000   
SUKIN Signature Original Hydrating Mist Toner                   0.000000   
Innisfree Jeju Cherry Blossom Mist 120 mL                       0.000000   
AVOSKIN YOUR SKIN BAE SERIES Panthenol 5% + Mug...       

# Getting Recommmendation


In [ ]:
def skincare_recommendation(product_name, similarity_data=cosine_sim_df, items=df[['product_name', 'brand', 'product_type', 'notable_effects', 'price']]):
    index = similarity_data.loc[:,product_name].to_numpy().argpartition(range(-1,-k,-1))
    #retriving data with greatest similarity
    closest=similarity_data.columns[index[-1:-(k+2):-1]]
    closest=closest.drop(product_name,errors='ignore')
    # Drop a name so that the name of the product we are looking for doesnt' appear in the list of recommendations
    closest=closest.drop(product_name,errors='ignore')
    return pd.DataFrame(closest).merge(items).head(k)

In [ ]:
df[df.product_name.eq('ELSHE SKIN Radiant Supple Serum')].head()


product_href  \
292  https://www.beautyhaul.com/product/detail/radi...   

                        product_name product_type       brand  \
292  ELSHE SKIN Radiant Supple Serum        Serum  ELSHE SKIN   

                        notable_effects       price  \
292  Acne-Free, Brightening, Black-Spot  Rp 182.750   

                                           description  \
292  ElsheSkin Radiant Supple Serum atau Radiant Sk...   

                                           picture_src  Combination  Dry  \
292  https://www.beautyhaul.com/assets/uploads/prod...            1    1   

     Normal  Oily  Sensitive  
292       1     1          1

In [ ]:
def skincare_recommendation(product_name, k, similarity_data=cosine_sim_df, items=df[['product_name', 'brand', 'product_type', 'notable_effects', 'price']]):
    index = similarity_data.loc[:,product_name].to_numpy().argpartition(range(-1,-k,-1))
    #retriving data with greatest similarity
    closest=similarity_data.columns[index[-1:-(k+2):-1]]
    closest=closest.drop(product_name,errors='ignore')
    # Drop a name so that the name of the product we are looking for doesnt' appear in the list of recommendations
    closest=closest.drop(product_name,errors='ignore')
    return pd.DataFrame(closest).merge(items).head(k)
skincare_recommendation('ELSHE SKIN Radiant Supple Serum', k=5)

product_name             brand  \
0                SOMETHINC Dark Spot Reducer Ampoule         SOMETHINC   
1  AVOSKIN YOUR SKIN BAE SERIES Niacinamide 12% +...           AVOSKIN   
2  Mineral Botanica Perfect Purifying Acne Night ...  MINERAL BOTANICA   
3  Wardah Crystal Secret Dark Spot & Brightening ...            WARDAH   
4       Laneige LNG White Dew Emulsion 100ml (OL21)            LANEIGE   

  product_type                                notable_effects       price  
0        Serum             Acne-Free, Brightening, Black-Spot  Rp 129.000  
1        Serum  Hydrating, Acne-Free, Brightening, Black-Spot  Rp 125.100  
2  Moisturizer            Moisturizing, Acne-Free, Black-Spot   Rp 52.000  
3        Serum          Moisturizing, Brightening, Black-Spot   Rp 60.100  
4  Moisturizer          Moisturizing, Brightening, Black-Spot  Rp 566.714

In [ ]:
df[df.product_name.eq('SOMETHINC Dark Spot Reducer Ampoule')].head()

product_href  \
1029  https://www.beautyhaul.com/product/detail/dark...   

                             product_name product_type      brand  \
1029  SOMETHINC Dark Spot Reducer Ampoule        Serum  SOMETHINC   

                         notable_effects       price  \
1029  Acne-Free, Brightening, Black-Spot  Rp 129.000   

                                            description  \
1029  Solusi bekas jerawat menghitam, flek, dan hipe...   

                                            picture_src  Combination  Dry  \
1029  https://www.beautyhaul.com/assets/uploads/prod...            0    1   

      Normal  Oily  Sensitive  
1029       0     0          1

In [ ]:
skincare_recommendation('SOMETHINC Dark Spot Reducer Ampoule', k=5)

product_name             brand  \
0                    ELSHE SKIN Radiant Supple Serum        ELSHE SKIN   
1  AVOSKIN YOUR SKIN BAE SERIES Niacinamide 12% +...           AVOSKIN   
2  Mineral Botanica Perfect Purifying Acne Night ...  MINERAL BOTANICA   
3  Wardah Crystal Secret Dark Spot & Brightening ...            WARDAH   
4       Laneige LNG White Dew Emulsion 100ml (OL21)            LANEIGE   

  product_type                                notable_effects       price  
0        Serum             Acne-Free, Brightening, Black-Spot  Rp 182.750  
1        Serum  Hydrating, Acne-Free, Brightening, Black-Spot  Rp 125.100  
2  Moisturizer            Moisturizing, Acne-Free, Black-Spot   Rp 52.000  
3        Serum          Moisturizing, Brightening, Black-Spot   Rp 60.100  
4  Moisturizer          Moisturizing, Brightening, Black-Spot  Rp 566.714